test_semver.py

```diff
@@ -505,3 +505,19 @@ def test_should_compare_prerelease_and_build_with_numbers():
     assert VersionInfo(1, 9, 1, 1, 1) < VersionInfo(1, 9, 1, 2, 1)
     assert VersionInfo('2') < VersionInfo(10)
     assert VersionInfo('2') < VersionInfo('10')
+
+
+def test_should_be_able_to_use_strings_as_major_minor_patch():
+    v = VersionInfo('1', '2', '3')
+    assert isinstance(v.major, int)
+    assert isinstance(v.minor, int)
+    assert isinstance(v.patch, int)
+    assert v.prerelease is None
+    assert v.build is None
+
+
+def test_should_be_able_to_use_integers_as_prerelease_build():
+    v = VersionInfo('1', '2', '3', 4, 5)
+    assert isinstance(v.prerelease, str)
+    assert isinstance(v.build, str)
+    assert VersionInfo('1', '2', '3') == VersionInfo(1, 2, 3)
```

Comment:
Move this to `test_should_be_able_to_use_strings_as_major_minor_patch` and add here: `assert VersionInfo(1, 2, 3, 4, 5) == VersionInfo(1, 2, 3, '4', '5')`. Also use integers in the ctor at line 520 as this is about the prerelease and build members to be "unconventional" types not the first 3.

Reply:
> Move this to `test_should_be_able_to_use_strings_as_major_minor_patch` and add here: `assert VersionInfo(1, 2, 3, 4, 5) == VersionInfo(1, 2, 3, '4', '5')`. 

Ok, I did this in commit 9361751. However, the term "here" is ambiguous. :wink: Better check the commit.

> Also use integers in the ctor at line 520 as this is about the prerelease and build members to be "unconventional" types not the first 3.

Not sure if I got it. The `VersionInfo.__init__` already have an `int()` and `str()` call. Maybe show it with an example what you mean?

Reply:
My suggestion would be (fix the '???'):

```
def test_should_be_able_to_use_strings_as_major_minor_patch():
    v = VersionInfo('1', '2', '3')
    assert isinstance(v.major, int)
    assert isinstance(v.minor, int)
    assert isinstance(v.patch, int)
    assert v.prerelease is None
    assert v.build is None
    assert VersionInfo('1', '2', '3') == VersionInfo(1, 2, 3)


def test_using_non_numeric_string_as_major_minor_patch_throws():
    with pytest.raises('???'):
        v = VersionInfo('a')
    with pytest.raises('???'):
        v = VersionInfo(1, 'a')
    with pytest.raises('???'):
        v = VersionInfo(1, 2, 'a')


def test_should_be_able_to_use_integers_as_prerelease_build():
    v = VersionInfo(1, 2, 3, 4, 5)
    assert isinstance(v.prerelease, str)
    assert isinstance(v.build, str)
    assert VersionInfo(1, 2, 3, 4, 5) == VersionInfo(1, 2, 3, '4', '5')
```

The last line in the changes tests is very explicit on purpose to serve as easily readable documentation. Important change is the focus on what is tested in each test and addition of a "corner case" test

Reply:
Thanks for the code, very appreciated! I've covered the `???`, it's a `ValueError` exception. I've copied and adapted your code in commit 9361751. Many thanks! :+1: